[cocoa] Support setting partitioned cookies from JS and storing partitioned cookies in cookie cache

sysrqb opened the pull request:
#### 16daa23d3ea0022332713096723573f218e149d8
<pre>
[cocoa] Support setting partitioned cookies from JS and storing partitioned cookies in cookie cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=284550">https://bugs.webkit.org/show_bug.cgi?id=284550</a>
<a href="https://rdar.apple.com/141358106">rdar://141358106</a>

Reviewed by Sihui Liu.

This patch:
 - sets the appropriate partition key when setting a cookie from JavaScript
 - enables partitioned cookies in the in-process cookie cache
 - extends NetworkStorageSession::domCookiesForHost so it returns first-party partitioned cookies

Covered by new layout test. The expected result shows that the implementation
does not currently support two cookies with the same name where one is
partitioned and the other is not.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/cookies/multiple-cookies-with-partitioned.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/multiple-cookies-with-partitioned.https.html: Added.
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::domCookiesForHost):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::parseDOMCookie):
(WebCore::NetworkStorageSession::setCookiesFromDOM const):
(WebCore::NetworkStorageSession::domCookiesForHost):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::domCookiesForHost):
* Source/WebKit/Shared/WebProcessDataStoreParameters.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::preferencesDidChange):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::setOptInCookiePartitioningEnabled):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::propagateSettingUpdates):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieCacheCocoa.mm:
(WebKit::WebCookieCache::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setWebsiteDataStoreParameters):
(WebKit::WebProcess::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: <a href="https://commits.webkit.org/288727@main">https://commits.webkit.org/288727@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a548b2640db1b24d980b6eb3fca47164f2102e08

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84203 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38507 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89279 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35212 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86288 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3912 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11797 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65500 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23336 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87249 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2933 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76516 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45793 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2881 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30746 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34261 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73797 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31510 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90660 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11470 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8327 "Found 1 new test failure: fast/selectors/selection-window-inactive-stroke-color.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73955 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11696 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72341 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73157 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17463 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2835 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13032 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11422 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16898 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11271 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14747 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13044 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -434,7 +434,7 @@ - (CFURLRef)_cfurl;
 }
 
-static NSHTTPCookie *parseDOMCookie(String cookieString, NSURL* cookieURL, std::optional<Seconds> cappedLifetime)
+static NSHTTPCookie *parseDOMCookie(String cookieString, NSURL* cookieURL, std::optional<Seconds> cappedLifetime, NSString* partition = nil)
```
We probably should use `const String&` here and `nsStringNilIfEmpty` when passing it to `_cookieForSetCookieString` below; otherwise we should use `RetainPtr`.

szewai commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -856,6 +856,10 @@ void WebProcessProxy::addExistingWebPage(WebPageProxy& webPage, BeginsUsingDataS
     protectedWebsiteDataStore()->propagateSettingUpdatesToNetworkProcess();
+#if HAVE(ALLOW_ONLY_PARTITIONED_COOKIES)
+    if (auto [didChange, enabled] = protectedWebsiteDataStore()->didOptInCookiePartitioningStateChange(); didChange)
+        protectedProcessPool()->sendToAllProcesses(Messages::WebProcess::SetOptInCookiePartitioningEnabled(enabled));
```
Web content process needs to get `WebsiteDataStore::m_isOptInCookiePartitioningEnabled` value on initialization (likely in `WebProcess::SetWebsiteDataStoreParameters`), before UI process sends out message on update.

Also `WebsiteDataStore:: m_isOptInCookiePartitioningEnabled ` can be updated in 3 cases:
1. page is added
2. page is removed
3. page's `isOptInCookiePartitioningEnabled` preference value has changed

It seems we don't update the value in the third case.

sysrqb replied:
> Web content process needs to get WebsiteDataStore::m_isOptInCookiePartitioningEnabled value on initialization

This PR should already do that in `WebProcessPool::webProcessDataStoreParameters`. That is called from `WebProcessPool::initializeNewWebProcess`.

> 3. page's isOptInCookiePartitioningEnabled preference value has changed

I intentionally ignored this case for simplicity, it's the same behavior as the blob partitioning flag. But adding a check in `WebPageProxy::preferencesDidChange` may be sufficient? I'll test that.

szewai reviewed (commented):
I think partitioned (3rd-party) cookies are not eligible for `WebCookieCache` in existing implementation, see `WebCookieJar::isEligibleForCache`, so the changes about cookie cache are likely no-op (the one in `NetworkStorageSessionCocoa` has).

To add the support you will need to:
1. Actually add partitioned cookies to cache
2. Update cached partitioned cookies when there is a change (need to check if CookieChangeListener works for partitioned cookies)

And I think we could add a test for them to validate, e.g. by adding window.internals API to check if cookie is in cache

szewai commented on Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
```diff
@@ -713,9 +719,47 @@ static NSHTTPCookieAcceptPolicy httpCookieAcceptPolicy(CFHTTPCookieStorageRef co
+            }
+        };
+
+        [nsCookieStorage() _getCookiesForPartition:(NSString *)partitionKey completionHandler:completionHandler];
```
If this is new SPI, I think you need to add it to CFNetworkSPI.h

sysrqb replied:
Thanks, missed that.

Commit:
[cocoa] Support setting partitioned cookies from JS and storing partitioned cookies in cookie cache
https://bugs.webkit.org/show_bug.cgi?id=284550
rdar://141358106

Reviewed by Sihui Liu.

This patch:
 - sets the appropriate partition key when setting a cookie from JavaScript
 - enables partitioned cookies in the in-process cookie cache
 - extends NetworkStorageSession::domCookiesForHost so it returns first-party partitioned cookies

Covered by new layout test. The expected result shows that the implementation
does not currently support two cookies with the same name where one is
partitioned and the other is not.

* LayoutTests/TestExpectations:
* LayoutTests/http/tests/cookies/multiple-cookies-with-partitioned.https-expected.txt: Added.
* LayoutTests/http/tests/cookies/multiple-cookies-with-partitioned.https.html: Added.
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::domCookiesForHost):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::parseDOMCookie):
(WebCore::NetworkStorageSession::setCookiesFromDOM const):
(WebCore::NetworkStorageSession::domCookiesForHost):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::domCookiesForHost):
* Source/WebKit/Shared/WebProcessDataStoreParameters.h:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::preferencesDidChange):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::webProcessDataStoreParameters):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::setOptInCookiePartitioningEnabled):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::propagateSettingUpdates):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieCacheCocoa.mm:
(WebKit::WebCookieCache::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
(WebKit::WebCookieCache::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebPage/WebCookieCache.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setWebsiteDataStoreParameters):
(WebKit::WebProcess::setOptInCookiePartitioningEnabled):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: https://commits.webkit.org/288727@main
